bam/reader.go

```diff
@@ -109,90 +160,63 @@ func (br *Reader) Read() (*sam.Record, error) {
 		return nil, io.EOF
 	}
 
-	r := errReader{r: br.r}
-	bin := binaryReader{r: &r}
-
-	// Read record header data.
-	blockSize := int(bin.readInt32())
-	r.n = 0 // The blocksize field is not included in the blocksize.
-
+	var buf [4]byte
+	n, err := br.r.Read(buf[:4])
+	if err != nil {
+		return nil, err
+	}
+	if n != 4 {
+		return nil, errors.New("bam: invalid record: short block size")
+	}
 	// br.r.Chunk() is only valid after the call the Read(), so this
 	// must come after the first read in the record.
 	tx := br.r.Begin()
 	defer func() {
 		br.lastChunk = tx.End()
 	}()
 
-	var rec sam.Record
-
-	refID := bin.readInt32()
-	rec.Pos = int(bin.readInt32())
-	nLen := bin.readUint8()
-	rec.MapQ = bin.readUint8()
-	_ = bin.readUint16()
-	nCigar := bin.readUint16()
-	rec.Flags = sam.Flags(bin.readUint16())
-	lSeq := bin.readInt32()
-	nextRefID := bin.readInt32()
-	rec.MatePos = int(bin.readInt32())
-	rec.TempLen = int(bin.readInt32())
-	if r.err != nil {
-		return nil, r.err
+	b, err := newBuffer(br.r, buf)
```

Comment:
I was wrong before about returning the bgzf.Tx. It is even better. Since the entire record has been read, the lastChunk field can be populated in newBuffer, eliding the need for the defer.